### PR TITLE
OCaml API 0.3.0: Updated P-256 interface and build fixes

### PR DESCRIPTION
**Pull request by victor-dumitrescu**
Updates to the P-256 API have been merged as part of #331. This PR contains a simplification of the hacl-star-raw build process along with a fix to an issue encountered in some configurations missing a path when building OCaml bytecode that uses the library.

### CODE REVIEW

File: dist/Makefile.tmpl [hunk 1]
```diff
@@ -144,8 +144,6 @@ lib_gen/Lib_RandomBuffer_System_gen.exe: lib/Lib_RandomBuffer_System_bindings.cm
 # https://www.gnu.org/software/make/manual/html_node/Canceling-Rules.html
 %.exe:
 
-all: ocamlevercrypt.cmxa
```
Comment (msprotz) [inline dist/Makefile.tmpl:147]: why remove this?

File: dist/Makefile.tmpl [hunk 2]
```diff
@@ -180,18 +178,24 @@ CTYPES_CMX    =  $(filter-out $(BLACKLIST_CMX),$(CTYPES_DEPS))
...
-	ocamlfind mklib -o ocamlevercrypt $(CTYPES_CMX) $(ALL_C_STUBS) -ccopt -L. -cclib -levercrypt
+	ocamlmklib -o ocamlevercrypt $(CTYPES_CMX) -L. -levercrypt
 
-CTYPES_CMO    = $(patsubst %.cmx,%.cmo,$(CTYPES_CMX))
-ALL_ML_STUBS_CMO = $(patsubst %.cmx,%.cmo,$(ALL_ML_STUBS))
+STUBLIBS_PATH=$(OPAM_SWITCH_PREFIX)/lib/stublibs
 
-ocamlevercrypt.cma: $(ALL_BINDINGS) $(ALL_ML_STUBS_CMO) $(CTYPES_CMO) $(ALL_C_STUBS) libevercrypt.$(SO)
-	ocamlfind mklib -o ocamlevercrypt $(CTYPES_CMO) $(ALL_C_STUBS) -ccopt -L. -cclib -levercrypt
+dllocamlevercrypt.so: ocamlevercrypt.cmxa ocamlevercrypt.cma
```
Comment (msprotz) [inline dist/Makefile.tmpl:192]: this is `.dll` on windows; please use `$(SO)`

File: dist/Makefile.tmpl [hunk 3]
```diff
@@ -180,18 +178,24 @@ CTYPES_CMX    =  $(filter-out $(BLACKLIST_CMX),$(CTYPES_DEPS))
...
 CTYPES_CMI    =  $(patsubst %.cmx,%.cmi,$(CTYPES_CMX))
+CTYPES_CMO    = $(patsubst %.cmx,%.cmo,$(CTYPES_CMX))
+
+
+ocamlevercrypt.cma: $(ALL_BINDINGS) $(CTYPES_CMO) $(ALL_C_STUBS) $(CTYPES_CMX) libevercrypt.$(SO)
+	ocamlmklib -o ocamlevercrypt $(CTYPES_CMO) -L. -levercrypt
 
 ocamlevercrypt.cmxa: $(ALL_BINDINGS) $(ALL_ML_STUBS) $(ALL_C_STUBS) libevercrypt.$(SO)
-	ocamlfind mklib -o ocamlevercrypt $(CTYPES_CMX) $(ALL_C_STUBS) -ccopt -L. -cclib -levercrypt
+	ocamlmklib -o ocamlevercrypt $(CTYPES_CMX) -L. -levercrypt
```
Comment (msprotz) [inline dist/Makefile.tmpl:188]: why not prefix with `ocamlfind`?

**Review by msprotz — COMMENTED** [review]
Looks good, I have a few minor questions... thanks!

**Review by msprotz — COMMENTED** [review]
Looks good... minus a few comments. Thanks!

File: dist/Makefile.tmpl (hunk 1 above)
Comment (victor-dumitrescu) [reply dist/Makefile.tmpl:147]: This was redundant, there's an earlier `all: libevercrypt.$(SO)`

File: dist/Makefile.tmpl (hunk 3 above)
Comment (victor-dumitrescu) [reply dist/Makefile.tmpl:188]: I followed the example in the docs, but the result seems to be the same

File: dist/Makefile.tmpl (hunk 1 above)
Comment (msprotz) [reply dist/Makefile.tmpl:147]: I don't understand,building libevercrypt.$(SO) doesn't imply building `ocamlevercrypt.cmxa`, does it?

Comment (victor-dumitrescu) [reply dist/Makefile.tmpl:147]: No, but I don't think this changes anything, this target was basically dead, right? `ocamlevercrypt.cmxa` is built by `install-hacl-star-raw`.

Comment (victor-dumitrescu) [reply dist/Makefile.tmpl:147]: Fair enough, but can that work is the Makefile is shared but we're only building the OCaml library in gcc-compatible?

Comment (msprotz) [reply dist/Makefile.tmpl:147]: there's a guard here: 

103 ifneq (,$(wildcard lib_gen))

**Comment by msprotz** [comment]
looks good to me thanks!

**Review by msprotz — APPROVED** [review]

**Review by polubelova — APPROVED** [review]

**Comment by msprotz** [comment]
@victor-dumitrescu you should be able to merge this now as soon as it comes back green

**Pull request merged**